File: fix_quotes.js
const fs = require('fs');

// Read the HTML file
const htmlContent = fs.readFileSync('index.html', 'utf8');

// Find the testData section
const start = htmlContent.indexOf('window.testData = {');
const consoleLogPos = htmlContent.indexOf("console.log('window.testData loaded");
const testDataSection = htmlContent.substring(start, consoleLogPos);

// Extract just the object part
const objectStart = 'window.testData = '.length;
const objectStr = testDataSection.substring(objectStart);

// Function to convert single quotes to double quotes, but be careful about quotes inside strings
function convertQuotes(str) {
    let result = '';
    let inString = false;
    let stringChar = '';
    let escapeNext = false;

    for (let i = 0; i < str.length; i++) {
        const char = str[i];

        if (escapeNext) {
            result += char;
            escapeNext = false;
            continue;
        }

        if (char === '\\') {
            result += char;
            escapeNext = true;
            continue;
        }

        if (!inString) {
            if (char === '"' || char === "'") {
                inString = true;
                stringChar = char;
                result += '"'; // Convert to double quote
            } else {
                result += char;
            }
        } else {
            if (char === stringChar) {
                inString = false;
                result += '"'; // Convert closing quote to double quote
            } else {
                result += char;
            }
        }
    }

    return result;
}

// Convert quotes in the object
const convertedObject = convertQuotes(objectStr);

// Replace the testData section
const newTestDataSection = 'window.testData = ' + convertedObject;
const newHtmlContent = htmlContent.replace(testDataSection, newTestDataSection);

// Write back
fs.writeFileSync('index.html', newHtmlContent);
console.log('Converted all quotes to double quotes in testData');

File: fix_quotes.py
from __future__ import annotations

import re
from typing import Match

# Read the file
with open('index.html', 'r', encoding='utf-8') as f:
    content = f.read()

# Find testData section
start = content.find('window.testData = {')
end = content.find("console.log('window.testData loaded")
before_testdata = content[:start + len('window.testData = ')]
after_testdata = content[end:]
testdata_str = content[start + len('window.testData = '):end]

print(f'Original testData length: {len(testdata_str)}')

# Function to escape quotes inside JSON string values
def escape_quotes_in_values(match: Match[str]) -> str:
    key = match.group(1)
    value = match.group(2)
    # Escape any unescaped quotes in the value
    escaped_value = value.replace('\\"', '___ESCAPED___').replace('"', '\\"').replace('___ESCAPED___', '\\"')
    return f'"{key}": "{escaped_value}"'

# Pattern to match key-value pairs
# This finds patterns like: "key": "value with "quotes" inside"
pattern = r'"([^"]+)":\s*"([^"]*(?:"[^"]*)*)"(?=[,\n\r])'

# Apply fixes
fixed_testdata = testdata_str

# Simple approach: find all values with embedded quotes and fix them
lines: list[str] = testdata_str.split('\n')
fixed_lines: list[str] = []

for line in lines:
    # Check if line contains a string value with embedded quotes
    # Pattern: "key": "value with "embedded" quotes"
    if '": "' in line:
        # Find the value part
        match = re.search(r'("(?:[^"\\]|\\.)*"):\s*("(?:[^"\\]|\\.)+")', line)
        if match:
            key = match.group(1)
            value = match.group(2)
            # Check if value has unescaped quotes
            if '"' in value[1:-1] and '\\"' not in value:
                # Fix by escaping quotes in value
                inner_value = value[1:-1]  # Remove outer quotes
                escaped_inner = inner_value.replace('"', '\\"')
                fixed_value = f'"{escaped_inner}"'
                new_line = line.replace(value, fixed_value)
                fixed_lines.append(new_line)
                if line != new_line:
                    print(f'Fixed line: {line[:80]}...')
                continue
    
    fixed_lines.append(line)

fixed_testdata = '\n'.join(fixed_lines)

print(f'New testData length: {len(fixed_testdata)}')

# Reconstruct
new_content = before_testdata + fixed_testdata + after_testdata

# Write back
with open('index.html', 'w', encoding='utf-8') as f:
    f.write(new_content)

print('✓ Fixed embedded quotes')
